src/tmpfiles/tmpfiles.c

```diff
@@ -2839,6 +2840,39 @@ static int create_item(Context *c, Item *i) {
         return 0;
 }
 
+static int purge_item_instance(Context *c, Item *i, const char *instance, CreationMode creation) {
+        int r;
+
+        /* FIXME: we probably should use dir_cleanup() here instead of rm_rf() so that 'x' is honoured. */
+        log_debug("rm -rf \"%s\"", instance);
+        r = rm_rf(instance, REMOVE_ROOT|REMOVE_SUBVOLUME|REMOVE_PHYSICAL);
+        if (r < 0 && r != -ENOENT)
+                return log_error_errno(r, "rm_rf(%s): %m", instance);
+
+        return 0;
+}
+
+static int purge_item(Context *c, Item *i) {
+
+        assert(i);
+
+        log_debug("Running purge owned action for entry %c %s", (char) i->type, i->path);
+
+        if (IN_SET(i->type,
+                   COPY_FILES,
+                   TRUNCATE_FILE,
+                   CREATE_FILE,
+                   WRITE_FILE,
+                   EMPTY_DIRECTORY,
+                   CREATE_SYMLINK,
+                   CREATE_FIFO,
+                   CREATE_DIRECTORY,
+                   TRUNCATE_DIRECTORY))
+                return glob_item(c, i, purge_item_instance);
```

Comment:
hmm, this should only do glob_item() for the types that pass needs_glob()

Reply:
It stops working after making that change, so I've left it as-is

Reply:
why? please elaborate. We only do globbing for some line types, not for others. We should do this on all operations, also on purge

Reply:
Er, you tell me :-) It just doesn't work, try cloning the branch, adding `&& needs_glob(i->type)` and then running the simple test:

```
systemd-tmpfiles --create - <<EOF
d /tmp/somedir
f /tmp/somedir/somefile - - - - baz
EOF
```

and

```
systemd-tmpfiles --purge - <<EOF
d /tmp/somedir
f /tmp/somedir/somefile - - - - baz
EOF
```

and the file/dir are not deleted anymore as expected...

Reply:
Other uses of glob_item are also not behind needs_glob, like clean_item() and remove_item() which is what I copied it from - are you sure it isn't simply a case of a badly named function?

Reply:
> Other uses of glob_item are also not behind needs_glob, like clean_item() and remove_item() which is what I copied it from - are you sure it isn't simply a case of a badly named function?

@poettering ping - I don't know what else I can do here

Reply:
well, for the non-glob cases you should call purge_item_instance() directly, did you do this?

i.e.:


```c
if (!needs_purge(i->type))
        return 0;

if (needs_glob(i->type))
        return glob_item(c, i, purge_item_instance);

return purge_item_instance(c, i, i->path, CREATION_EXISTING);
```

See how clean_item() does this. For the globbing cases it goes via glob_item(), and otherwise calls clean_item_instance() directly.

Here we do the exact same.

Reply:
> Other uses of glob_item are also not behind needs_glob, like clean_item() and remove_item() which is what I copied it from - are you sure it isn't simply a case of a badly named function?

They use a switch statement locally, since they operate only some specific conditions, and sometimes slightly differently. But if you look at the switch cases they also do the glob_item() thing in all cases that needs_glob() returns true, and go directly in the others. 

I mean, i'd be OK to do a local switch for the purge case too to cover for the cases, that's up to you.

(Well, there's one minor bug in the current logic, will post a fix shortly, but it's minor)

Reply:
Ah ok, got confused by that - fixed as suggested, thanks